### PR TITLE
Event Device Support (`evdev`)

### PR DESCRIPTION
**Pull request by votti**
## Background
As described in  #1942, a plugin to support to setup listeners to event devices such as USB controller, joysticks or keyboards would be useful.

## Implementation
This implementation builds on the `EvDevKeyListener` and is based on the `bluetooth_audio_buttons` plugin.

The following details differ from the `bluetooth_audio_buttons` plugin:
- initialize the listeners from the configuration. I have tried to document the format in the docstring. Is there a better place to do so?
- rpc command is bound with `deference=False`. Otherwise plugins, eg next_song, may not be initialized yet upon initialization. Is there a better way?

### CODE REVIEW

**Comment by s-martin** [comment]
@pabera @ChisSoc

**Comment by ChisSoc** [comment]
Nice addition!

Regarding the documentation: The best way to go would be to write a Ueer Guide Page. The live in `docs/sphinx/userguide`. Just create a new `.rst` file and add to in the `index.rst`. See the bluetooth doc [as an example](https://rpi-jukebox-rfid.readthedocs.io/en/latest/userguide/bluetooth_audio_buttons.html). 

The user guide needs
* How to activate this plugin
* How to find out the device_name that is needed for the configuration
* How to find out the button id
* And a complete configuration example, please

Ah, yes, `dereference`. If I remember correctly this has to do with the order of the loading of the plugins. Try loading your plugin last.

**Comment by votti** [comment]
Thanks for the feedback+

Top, I will look into adding the documentation.

@dereference: i already have put the plugin last, but the only way to get
it to run was to add the listeners only when adding them as a "finalize"
step.
Is the performance hit without dereference really bad enough to worry about
this optimization?

Cheers!

**Comment by ChisSoc** [comment]
Ah yes, finalize it was. 

Don't worry about the performance impact. It is negligible.

**Comment by votti** [comment]
I added now a documentation.
It would be great if someone could test it. As a device eg an USB mouse works well.

**Review by pabera — COMMENTED** [review]
I read through your documentation. I found a few typos which I corrected myself rather than calling them out. I committed them to the same branch. I'd like @ChisSoc to do a last check on the Python code.

Great addition. (all though I wonder how my 3 year old would use a USB mouse to rule their Phoniebox 🤔)

**Comment by votti** [comment]
Thanks for the comments, I will address them!

`(all though I wonder how my 3 year old would use a USB mouse to rule their Phoniebox thinking)`

The USB mouse was more a very general illustration that this works with any USB device, not necessarily an endorsement :) 
The beter usecase for this is to use a cheap Joystick controller board in order to hook up arcade buttons in via USB in case the GPIO pins are occupied otherwise.
I am using: https://de.aliexpress.com/item/1005001604578101.html?spm=a2g0o.order_list.order_list_main.5.26875c5fq1PFcA&gatewayAdapt=glo2deu

**Comment by votti** [comment]
It seems that flake8 is failing due to flake8 violations in code not touched by this PR.
Schould I do a separare PR fixing all the pre-existing violations or rather fold the fixes into this PR?

**Comment by pabera** [comment]
> Schould I do a separare PR fixing all the pre-existing violations or rather fold the fixes into this PR?

A separate PR would be ok.

Would you still be interested in merging this?

**Comment by AlvinSchiller** [comment]
> It seems that flake8 is failing due to flake8 violations in code not touched by this PR. Schould I do a separare PR fixing all the pre-existing violations or rather fold the fixes into this PR?

For the fixes see https://github.com/MiczFlor/RPi-Jukebox-RFID/pull/1989

**Comment by votti** [comment]
Hi @pabera ,

Sorry for being unresponsive, I had quite lot of personal events that diminished my free time.

I rebased the PR to the latest dev branch, so the flake8 should be fixed and the cosmetic issues should be addressed.

**Comment by pabera** [comment]
Still seem to have some issues. Have you tried locally?

**Comment by pabera** [comment]
I have fixed the flake8 issue in another PR. If you pull all changes again, you should be able to pass flake8 properly

**Comment by votti** [comment]
Hi there, 
I rebased again and run `flake8` locally - now all looks clean on my side.

**Comment by votti** [comment]
I also found by chance that in the old Jukebox there was already an implementation/reference about this: https://github.com/MiczFlor/RPi-Jukebox-RFID/blob/develop/components/controls/buttons_usb_encoder/README.md?plain=1
Should I link this in the readme?

**Comment by s-martin** [comment]
@votti would you mind merging latest changes from `future3/develop`? Biggest change is probably that docs are now markdown.

I think then we could merge this PR

**Comment by votti** [comment]
Please have a look at my updates.
I am using this now since a year with a `Zero delay USB encoder` and some arcade buttons and it works well.

File: documentation/builders/event-devices.md [hunk 1]
```diff
@@ -0,0 +1,134 @@
+# Event devices
+
+## Background
+Event devices are generic input devices that are exposed in ``/dev/input``.
+This includes USB peripherals (Keyboards, Controllers, Joysticks or Mouse) as well as potentially bluetooth devices.
+
+A specific usecase for this could be, if a Zero Delay Arcade USB Encoder is used to wire arcade buttons instead of using GPIO pins.
+
```
Comment (s-martin) [inline documentation/builders/event-devices.md:8]: I would not render the Markdown file as plain.

File: documentation/builders/event-devices.md [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+```
+
+In this example output, the `{key-code}` would be `297`
+
+Alternatively, the device could also be setup without a mapping.
+Afterwards, when pressing keys, the key codes can be found in the log files. Press various buttons on your device,
+while watching the logs with `tail -f shared/logs/app.log`.
+Look for entries like `No callback registered for button ...`.
+
+### Specifying the `{rpc_command_defintion}`
```
Comment (s-martin) [inline documentation/builders/event-devices.md:96]: Typo. An "i" is missing.

File: src/jukebox/components/controls/event_devices/README.rst [hunk 3]
```diff
@@ -0,0 +1,131 @@
+Event devices are generic input devices that are exposed in ``/dev/input``.
```
Comment (s-martin) [inline src/jukebox/components/controls/event_devices/README.rst:1]: This file could be removed? You put the markdown version under documentation.

**Review by s-martin — APPROVED** [review]

File: documentation/builders/event-devices.md [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+
+A specific usecase for this could be, if a Zero Delay Arcade USB Encoder is used to wire arcade buttons instead of using GPIO pins.
+
+This functionality was previously implemented under the name of [USB buttons](https://github.com/MiczFlor/RPi-Jukebox-RFID/blob/develop/components/controls/buttons_usb_encoder/README.md?plain=1).
+
+The devices and their button mappings need to be mapped in the configuration file.
+
+## Configuration
+
+To configure event devices, first add the plugin as the last entry to the module list:
```
Comment (pabera) [inline documentation/builders/event-devices.md:15]: Will it always have to be the last? Why?

File: documentation/builders/event-devices.md [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+              {rpc_command_definition} # eg `alias: toggle`
+```
+The `{device nickname}` is only for your own orientation and can be choosen freely.
+For each device you need to figure out the `{device_name}` and the `{event_id}` corresponding to key strokes, as indicated in the sections below.
+
+### Identifying the `{device_name}`
+
+The `{device_name}` can be identified using the following Python snippet:
+
+``` Python
```
Comment (pabera) [inline documentation/builders/event-devices.md:42]: In the spirit of run_*.py files, would it make sense to provide this code as a configurator? Or at least as a runable script so users don't have to write the code themselves?

File: documentation/builders/event-devices.md [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+
+To configure event devices, first add the plugin as the last entry to the module list:
+
+``` yaml
+  modules:
+    named:
+      event_devices: controls.event_devices # Add as last entry
+```
+
+Then add the following section to the configuration:
```
Comment (pabera) [inline documentation/builders/event-devices.md:23]: Similar to [gpio.yaml](https://github.com/MiczFlor/RPi-Jukebox-RFID/blob/future3/develop/resources/default-settings/gpio.example.yaml), I recommend to outsource this to it's own file, e.g. `evdev.yaml`. You could then also proovide an `evdev.example.yaml`

File: documentation/builders/event-devices.md [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+
+### Specifying the `{rpc_command_definition}`
+
+Each key looks like `{key-code}: {rpc_command_definition}`.
+The RPC command follows the regular RPC command rules as defined in the [following documentation](./rpc-commands.md).
+
+
+## Full example config
+
+Here is a complete configuration example for a USB Joystick controller:
```
Comment (pabera) [inline documentation/builders/event-devices.md:104]: I wonder, how close can we get this definition to GPIO? It'll be important to maintain parity of functionality between the two input types to reduce confusion with users.

I am referring to 1) the yaml layout (like for [GPIO](https://github.com/MiczFlor/RPi-Jukebox-RFID/blob/future3/develop/resources/default-settings/gpio.example.yaml)) and 2) the actions types such es "press", "short/long press", rotary actions etc.

Comment (votti) [reply documentation/builders/event-devices.md:104]: Aiming to harmonize the interface with GPIO seems valuable.

However I doubt that a 1:1 mapping of the config is feasible as the GPIO interface supports a single device with multiple pins, while EVDEV would be multiple devices with multiple pins.

`short/long press` does need additional logic (see discussion here: https://github.com/gvalkov/python-evdev/issues/70).
Things like rotation may be feasible as well, but would need more investigation. In particular it would be helpful to have a device at hand that has such events as a concrete use case.
Output (toggling LEDs, sounds) are also supported by EVDEV, but I dont have a device at hand to test it.

My proposal forward would be that I try to restructure the yaml that it follows the `gpio` interface more closely and maybe add stubs for `short/long press` with an `NotImplementedError`.
Once merged we may add follow up issues to investigate the future directions.

**Review by votti — COMMENTED** [review]
Thanks for the feedback!

To me it is important that we get the interace (`yml` config) right before merging, so I see the following todos for me:

- [x] Factor out the config as `evdev.yml`
- [x] Add `evdev.example.yaml`
- [x] Align the  `evdev.yml` structure with the `gpio.yml`
- [x] Investigate why I indicated that the `evdev` module should be listed last.

The following issues I would like to postpone as follow-up issues after merging:

- [ ] Add configurator script

File: documentation/builders/event-devices.md (hunk 1 above)
Comment (s-martin) [reply documentation/builders/event-devices.md:8]: The link has a `?plain=1` at the end. I don't think it's necessary, but it's not that important :)

Comment (pabera) [reply documentation/builders/event-devices.md:8]: I thank he was referring to the addition of `?plain=1` at the end of the URL (your comment on 15-Aug)

File: documentation/builders/event-devices.md (hunk 7 above)
Comment (pabera) [reply documentation/builders/event-devices.md:104]: Sounds like a good idea. If a full harmonization is not possible, I agree it's still valuable to get as close as possible and document the differences.

File: documentation/builders/event-devices.md (hunk 1 above)
Comment (votti) [reply documentation/builders/event-devices.md:8]: Ah sorry now I see it - sorry I was looking at the lines above and could not see anything.
Fixed now. Thanks for pointing this out.

**Comment by pabera** [comment]
@votti We are planning to integrate this PR in our next Release https://github.com/MiczFlor/RPi-Jukebox-RFID/pull/2211 which we want to release in the next couple of days. Do you think you will be able to work on the changes suggested or will you need more time? In the second case, we would postpone the release of this PR.

**Comment by votti** [comment]
@pabera I am away for holidays so no access to a pi. However I just managed to get the docker dev environment (including evdev devices) runing, so I may be able to work on it till the end of the week.

No hard feelings if this does not make the next release. I think having the configuration file in a final form  before merging it is really important.

**Comment by votti** [comment]
I now aligned the configuration structure with `gpio.yaml`.
In contrast to `gpio`, multiple devices each with their own `input/output_devices` are supported.
Note that currently only `input_devices`  the `Button` type with the `on_press` action is implemented.

While extending the config to such new functionalities is now easy, this code is still relying on the pre-existing `EventDeviceListener` code backend. Expanding this mechanisms to support more features (eg long press etc) will require significant changes on this backend. 

" Investigate why I indicated that the evdev module should be listed last." - I did not find a good reason after trying also different orders. Really not sure why I wrote it like this before.

**Comment by votti** [comment]
I now also added some tests and fixed some bugs/edge cases found during testing. 

One issue I encountered is that `zmq` is not installed for the Github CI testing.
First I tried to fix the Github CI but then decided to better just use `mocks` to make my tests independent of `zmq`.

I will add an issue that it would be nice to eventually have `zmq` also in our github action workflow used for testing.

File: resources/default-settings/evdev.example.yaml [hunk 8]
```diff
@@ -0,0 +1,59 @@
+devices: # A list of evdev devices each containing one or multiple input/output devices
+  joystick: # A nickname for a device
+    device_name: DragonRise Inc.   Generic   USB # Device name
```
Comment (pabera) [inline resources/default-settings/evdev.example.yaml:3]: Is this just a pretty name or actually relevant to access/call the device?

File: test/evdev/test_evdev_init.py [hunk 9]
```diff
@@ -0,0 +1,185 @@
+""" Tests for the evdev __init__ module
```
Comment (pabera) [inline test/evdev/test_evdev_init.py:1]: I like the fact that you have put effort into tests as well. It keeps pushing the bar. It's something we haven't pushed for in the past but indeed allows to be more confident when others make changes in the future.

**Review by s-martin — APPROVED** [review]

File: resources/default-settings/evdev.example.yaml (hunk 8 above)
Comment (votti) [reply resources/default-settings/evdev.example.yaml:3]: The 'device_name' is relevant as it is used to get the device number (in contrast to the nickname line 2 which is cosmetic).

The docs contain a section how to look it up: https://github.com/MiczFlor/RPi-Jukebox-RFID/blob/49839adb01019cb296d67dc29c06cd9a6af0bc62/documentation/builders/event-devices.md#identifying-the-device_name

**Comment by s-martin** [comment]
Could be merged?

**Comment by votti** [comment]
Let me know if there is anything still required from my side!
If not I would appreciate if this would be merged before it diverges again from the main develop/future3 branch.

**Review by pabera — APPROVED** [review]
Approved. 

Last question @votti .. How do you execute the tests in this scenario?
